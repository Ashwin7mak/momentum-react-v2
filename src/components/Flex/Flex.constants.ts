const CLASS_PREFIX = 'md-flex';

const ALIGN_CONTENT = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  spaceBetween: 'space-between',
  spaceAround: 'space-around',
  spaceEvenly: 'space-evenly',
  stretch: 'stretch',
  baseline: 'baseline',
};

const JUSTIFY_CONTENT = {
  ...ALIGN_CONTENT,
  left: 'left',
  right: 'right',
};

const ALIGN_ITEMS = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  stretch: 'stretch',
  baseline: 'baseline',
  selfStart: 'self-start',
  selfEnd: 'self-end',
};

const DIRECTION = {
  row: 'row',
  column: 'column',
  rowReverse: 'row-reverse',
  columnReverse: 'column-reverse',
};

const WRAP = {
  wrap: 'wrap',
  nowrap: 'nowrap',
  wrapReverse: 'wrap-reverse',
};

const DEFAULTS = {
  DIRECTION: DIRECTION.row,
  JUSTIFY_CONTENT: JUSTIFY_CONTENT.start,
  ALIGN_ITEMS: ALIGN_ITEMS.start,
  ALIGN_CONTENT: ALIGN_CONTENT.start,
  WRAP: WRAP.nowrap,
};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { DEFAULTS, STYLE, ALIGN_CONTENT, JUSTIFY_CONTENT, ALIGN_ITEMS, DIRECTION, WRAP };
