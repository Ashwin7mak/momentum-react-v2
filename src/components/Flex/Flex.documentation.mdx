`<Flex />` is a utility component that acts like an abstract layer which displays child components using flex.
