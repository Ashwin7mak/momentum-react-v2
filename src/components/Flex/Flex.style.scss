.md-flex-wrapper {
  &[data-xgap=true] {
    > *:not(:first-child) {
      margin-left: var(--flex-xgap-size);
    }
  }

  &[data-ygap=true] {
    > *:not(:first-child) {
      margin-top: var(--flex-ygap-size);
    }
  }
}
