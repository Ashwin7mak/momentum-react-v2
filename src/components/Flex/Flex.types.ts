import { ReactNode } from 'react';

export type FlexStyleProps = {
  /**
   * The distribution of space around items along the main axis. See [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/justify-content).
   * @default 'stretch'
   */
  justifyContent?:
    | 'start'
    | 'end'
    | 'center'
    | 'left'
    | 'right'
    | 'space-between'
    | 'space-around'
    | 'space-evenly'
    | 'stretch'
    | 'baseline';
  /**
   * The distribution of space around child items along the cross axis. See [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/align-content).
   * @default 'start'
   */
  alignContent?:
    | 'start'
    | 'end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly'
    | 'stretch'
    | 'baseline';
  /**
   * The alignment of children within their container. See [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/align-items).
   * @default 'stretch'
   */
  alignItems?: 'start' | 'end' | 'center' | 'stretch' | 'self-start' | 'self-end' | 'baseline';
  /**
   * The direction in which to layout children. See [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/flex-direction).
   * @default 'row'
   */
  direction?: 'row' | 'column' | 'row-reverse' | 'column-reverse';
  /**
   * Whether to wrap items onto multiple lines. See [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/flex-wrap).
   * @default false
   */
  wrap?: boolean | 'wrap' | 'nowrap' | 'wrap-reverse';

  /**
   * If present, it will add a margin-left to all elements except first child
   */
  xgap?: string;

  /**
   * If present, it will add a margin-top to all elements except first child
   */
  ygap?: string;
};

export interface Props extends FlexStyleProps {
  /**
   * className prop description
   */
  className?: string;
  children?: ReactNode;
}
