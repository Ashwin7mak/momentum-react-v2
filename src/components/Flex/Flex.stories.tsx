import React, { FC } from 'react';
import { Story } from '@storybook/react';

import Flex, { FlexProps } from './';
import {
  Title,
  Subtitle,
  Description,
  Primary,
  ArgsTable,
  PRIMARY_STORY,
} from '@storybook/addon-docs';

import Documentation from './Flex.documentation.mdx';
import argTypes from './Flex.stories.args';

const DocsPage: FC = () => (
  <>
    <Title />
    <Subtitle />
    <Description />
    <Documentation />
    <Primary />
    <ArgsTable story={PRIMARY_STORY} />
  </>
);

export default {
  title: 'Momentum UI/Flex',
  component: Flex,
  parameters: {
    expanded: true,
    docs: {
      page: DocsPage,
    },
  },
  argTypes: {
    ...argTypes,
  },
};

const Items = () => {
  return (
    <>
      <p style={{ border: '1px solid blue' }}>Item 1</p>
      <p style={{ border: '1px solid blue' }}>Item 2</p>
      <p style={{ border: '1px solid blue' }}>Item 3</p>
    </>
  );
};

const Template: Story<FlexProps> = (args: FlexProps) => (
  <Flex {...args} style={{ border: '1px solid red', minWidth: '20rem', height: '20rem' }}>
    <Items />
  </Flex>
);

const MultiTemplate: Story<FlexProps> = (args: FlexProps, { parameters }) => {
  const { variants } = parameters;

  const items = variants.map((variant, index: number) => (
    <div key={index}>
      <Flex {...args} {...variant} style={{ minWidth: '20rem', border: '1px solid red' }} />
      <p>{variant.label}</p>
    </div>
  ));

  return (
    <div
      style={{
        display: 'grid',
        gridTemplateColumns: `repeat(3, auto)`,
        gap: '1.5rem',
        alignItems: 'end',
      }}
    >
      {items}
    </div>
  );
};

const Example = Template.bind({});

const Common = MultiTemplate.bind({});

Common.parameters = {
  variants: [
    { children: <Items />, justifyContent: 'flex-start', xgap: '1rem' },
    { children: <Items />, justifyContent: 'center', xgap: '1rem' },
    { children: <Items />, justifyContent: 'flex-end', xgap: '1rem' },
    { children: <Items />, direction: 'column', alignItems: 'flex-start', ygap: '1rem' },
    { children: <Items />, direction: 'column', alignItems: 'center', ygap: '1rem' },
    { children: <Items />, direction: 'column', alignItems: 'flex-end', ygap: '1rem' },
  ],
};

export { Example, Common };
