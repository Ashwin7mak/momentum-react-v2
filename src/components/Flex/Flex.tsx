import React, { CSSProperties, FC } from 'react';
import classnames from 'classnames';

import './Flex.style.scss';
import { Props } from './Flex.types';
import { DEFAULTS, STYLE } from './Flex.constants';

const Flex: FC<Props> = (props: Props) => {
  const {
    className,
    children,
    xgap,
    ygap,
    direction,
    justifyContent,
    alignItems,
    alignContent,
    wrap,
    style,
    id,
    ...rest
  } = props;
  const _style = {
    flexDirection: direction || DEFAULTS.DIRECTION,
    justifyContent: justifyContent || DEFAULTS.JUSTIFY_CONTENT,
    alignItems: alignItems || DEFAULTS.ALIGN_ITEMS,
    alignContent: alignContent || DEFAULTS.ALIGN_CONTENT,
    flexWrap: wrap || 'nowrap',
  };
  return (
    <div
      {...rest}
      id={id}
      data-xgap={_style.flexDirection === 'row' && true}
      data-ygap={_style.flexDirection === 'column' && true}
      style={{
        display: 'flex',
        ...(_style as CSSProperties),
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        ['--flex-xgap-size' as any]: xgap,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        ['--flex-ygap-size' as any]: ygap,
        ...style,
      }}
      className={classnames(className, STYLE.wrapper)}
    >
      {children}
    </div>
  );
};

/**
 * TODO: Add description of component here.
 */

export default Flex;
