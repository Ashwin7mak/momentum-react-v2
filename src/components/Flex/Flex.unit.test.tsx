import Flex from '.';
import { mount } from 'enzyme';
import React from 'react';

describe('Flex', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      /* ...snapshot test... */
    });
  });

  describe('attributes', () => {
    it('should have X value', () => {
      /* ...attribute tests... */
    });
  });
});
