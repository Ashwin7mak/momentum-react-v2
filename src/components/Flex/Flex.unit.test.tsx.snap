// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex snapshot should match snapshot 1`] = `
<Flex>
  <div
    className="md-flex-wrapper"
    data-xgap={true}
    data-ygap={false}
    style={
      Object {
        "--flex-xgap-size": undefined,
        "--flex-ygap-size": undefined,
        "alignContent": "start",
        "alignItems": "start",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "start",
      }
    }
  >
    Test
  </div>
</Flex>
`;

exports[`Flex snapshot should match snapshot with alignContent 1`] = `
<Flex
  alignContent="center"
>
  <div
    className="md-flex-wrapper"
    data-xgap={true}
    data-ygap={false}
    style={
      Object {
        "--flex-xgap-size": undefined,
        "--flex-ygap-size": undefined,
        "alignContent": "center",
        "alignItems": "start",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "start",
      }
    }
  >
    Test
  </div>
</Flex>
`;

exports[`Flex snapshot should match snapshot with alignItems 1`] = `
<Flex
  alignItems="center"
>
  <div
    className="md-flex-wrapper"
    data-xgap={true}
    data-ygap={false}
    style={
      Object {
        "--flex-xgap-size": undefined,
        "--flex-ygap-size": undefined,
        "alignContent": "start",
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "start",
      }
    }
  >
    Test
  </div>
</Flex>
`;

exports[`Flex snapshot should match snapshot with className 1`] = `
<Flex
  className="example-class"
>
  <div
    className="example-class md-flex-wrapper"
    data-xgap={true}
    data-ygap={false}
    style={
      Object {
        "--flex-xgap-size": undefined,
        "--flex-ygap-size": undefined,
        "alignContent": "start",
        "alignItems": "start",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "start",
      }
    }
  >
    Test
  </div>
</Flex>
`;

exports[`Flex snapshot should match snapshot with id 1`] = `
<Flex
  id="example-id"
>
  <div
    className="md-flex-wrapper"
    data-xgap={true}
    data-ygap={false}
    id="example-id"
    style={
      Object {
        "--flex-xgap-size": undefined,
        "--flex-ygap-size": undefined,
        "alignContent": "start",
        "alignItems": "start",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "start",
      }
    }
  >
    Test
  </div>
</Flex>
`;

exports[`Flex snapshot should match snapshot with justifyContent 1`] = `
<Flex
  justifyContent="center"
>
  <div
    className="md-flex-wrapper"
    data-xgap={true}
    data-ygap={false}
    style={
      Object {
        "--flex-xgap-size": undefined,
        "--flex-ygap-size": undefined,
        "alignContent": "start",
        "alignItems": "start",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "center",
      }
    }
  >
    Test
  </div>
</Flex>
`;

exports[`Flex snapshot should match snapshot with style 1`] = `
<Flex
  style={
    Object {
      "color": "pink",
    }
  }
>
  <div
    className="md-flex-wrapper"
    data-xgap={true}
    data-ygap={false}
    style={
      Object {
        "--flex-xgap-size": undefined,
        "--flex-ygap-size": undefined,
        "alignContent": "start",
        "alignItems": "start",
        "color": "pink",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "start",
      }
    }
  >
    Test
  </div>
</Flex>
`;

exports[`Flex snapshot should match snapshot with wrap 1`] = `
<Flex
  wrap="wrap-reverse"
>
  <div
    className="md-flex-wrapper"
    data-xgap={true}
    data-ygap={false}
    style={
      Object {
        "--flex-xgap-size": undefined,
        "--flex-ygap-size": undefined,
        "alignContent": "start",
        "alignItems": "start",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "wrap-reverse",
        "justifyContent": "start",
      }
    }
  >
    Test
  </div>
</Flex>
`;

exports[`Flex snapshot should match snapshot with xgap 1`] = `
<Flex
  xgap="1rem"
>
  <div
    className="md-flex-wrapper"
    data-xgap={true}
    data-ygap={false}
    style={
      Object {
        "--flex-xgap-size": "1rem",
        "--flex-ygap-size": undefined,
        "alignContent": "start",
        "alignItems": "start",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "start",
      }
    }
  >
    Test
  </div>
</Flex>
`;

exports[`Flex snapshot should match snapshot with ygap 1`] = `
<Flex
  ygap="1rem"
>
  <div
    className="md-flex-wrapper"
    data-xgap={true}
    data-ygap={false}
    style={
      Object {
        "--flex-xgap-size": undefined,
        "--flex-ygap-size": "1rem",
        "alignContent": "start",
        "alignItems": "start",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "start",
      }
    }
  >
    Test
  </div>
</Flex>
`;
